### ISSUE TITLE
'Cairo error "out of memory"' when rendering SVG images

### ISSUE THREAD

**Issue by enstyled** [imgproxy/imgproxy#646]
We are running into an issue with some SVG files (attached below) when rendering to PNG/JPG/WEBP:
 - If the requested raster size is less than 200px, they render correctly
 - Above 200px, they render blank (white)
 - Above 420px, they crash imgproxy:

http://localhost:8080/insecure/width:500/height:500/plain/test/3505415.svg@png

When I enable`RUST_BACKTRACE=full`:
```
thread '<unnamed>' panicked at 'Cairo error "out of memory"', /root/deps/librsvg/vendor/cairo-rs/src/enums.rs:274:13
stack backtrace:
   0:     0x7f6e3cad57c0 - <unknown>
   1:     0x7f6e3cb505bc - <unknown>
   2:     0x7f6e3cac7322 - <unknown>
   3:     0x7f6e3cad9435 - <unknown>
   4:     0x7f6e3cad8f8a - <unknown>
   5:     0x7f6e3cad9d11 - <unknown>
   6:     0x7f6e3cad9857 - <unknown>
   7:     0x7f6e3cad5c7c - <unknown>
   8:     0x7f6e3cad97b9 - <unknown>
   9:     0x7f6e3cad976b - <unknown>
  10:     0x7f6e3c817a91 - <unknown>
  11:     0x7f6e3c76037c - <unknown>
  12:     0x7f6e3c77dc5d - <unknown>
  13:     0x7f6e3c666f81 - <unknown>
  14:     0x7f6e3c75b4f2 - <unknown>
  15:     0x7f6e3c66d782 - <unknown>
  16:     0x7f6e3c72367d - <unknown>
  17:     0x7f6e3c763238 - <unknown>
  18:     0x7f6e3c72396f - <unknown>
  19:     0x7f6e3c6809b5 - <unknown>
  20:     0x7f6e3c758a1a - <unknown>
  21:     0x7f6e3c680971 - <unknown>
  22:     0x7f6e3c666edc - <unknown>
  23:     0x7f6e3c75b4f2 - <unknown>
  24:     0x7f6e3c66d782 - <unknown>
  25:     0x7f6e3c72367d - <unknown>
  26:     0x7f6e3c763238 - <unknown>
  27:     0x7f6e3c72396f - <unknown>
  28:     0x7f6e3c6822f9 - <unknown>
  29:     0x7f6e3c758a1a - <unknown>
  30:     0x7f6e3c681ddb - <unknown>
  31:     0x7f6e3c666f31 - <unknown>
  32:     0x7f6e3c75b4f2 - <unknown>
  33:     0x7f6e3c66d782 - <unknown>
  34:     0x7f6e3c72367d - <unknown>
  35:     0x7f6e3c763238 - <unknown>
  36:     0x7f6e3c755b5d - <unknown>
  37:     0x7f6e3c767002 - <unknown>
  38:     0x7f6e3c60d915 - <unknown>
  39:     0x7f6e3c60fb07 - rsvg_rust_handle_render_cairo
  40:     0x7f6e3df7e3db - <unknown>
  41:     0x7f6e3dfd6cb1 - <unknown>
  42:     0x7f6e3dfd7d87 - <unknown>
  43:     0x7f6e3dfdba7e - vips_region_prepare_to
  44:     0x7f6e3df1893f - <unknown>
  45:     0x7f6e3dfd6cb1 - <unknown>
  46:     0x7f6e3dfd7d87 - <unknown>
  47:     0x7f6e3dfdba7e - vips_region_prepare_to
  48:     0x7f6e3df1893f - <unknown>
  49:     0x7f6e3dfd6cb1 - <unknown>
  50:     0x7f6e3dfd7855 - vips_region_fill
  51:     0x7f6e3dfdb8ad - vips_region_prepare
  52:     0x7f6e3df19a0b - <unknown>
  53:     0x7f6e3dfd6cb1 - <unknown>
  54:     0x7f6e3dfd7855 - vips_region_fill
  55:     0x7f6e3dfdb8ad - vips_region_prepare
  56:     0x7f6e3dfc96c9 - <unknown>
  57:     0x7f6e3dfd6cb1 - <unknown>
  58:     0x7f6e3dfd7855 - vips_region_fill
  59:     0x7f6e3dfdb8ad - vips_region_prepare
  60:     0x7f6e3df8b159 - <unknown>
  61:     0x7f6e3dfd6cb1 - <unknown>
  62:     0x7f6e3dfd7855 - vips_region_fill
  63:     0x7f6e3dfdb8ad - vips_region_prepare
  64:     0x7f6e3dfc96c9 - <unknown>
  65:     0x7f6e3dfd6cb1 - <unknown>
  66:     0x7f6e3dfd7d87 - <unknown>
  67:     0x7f6e3dfdba7e - vips_region_prepare_to
  68:     0x7f6e3dfd08c8 - <unknown>
  69:     0x7f6e3dfdc43a - <unknown>
  70:     0x7f6e3dfdc1bc - <unknown>
  71:     0x7f6e3dd070ad - <unknown>
  72:     0x7f6e3dc5efa3 - start_thread
  73:     0x7f6e3db754cf - clone
  74:                0x0 - <unknown>
fatal runtime error: failed to initiate panic, error 5
```

I have tried running the SVG files through [SVGO](https://github.com/svg/svgo), but they still crash imgproxy consistently.

[Archive.zip](https://github.com/imgproxy/imgproxy/files/6602259/Archive.zip)

**Comment by DarthSim** [comment]
Hey!
Sorry for the delay, I was on vacation without any access to GH.

It's a Cairo problem, and it's triggered by this: `patternTransform="matrix(1 0 0 -1 -120.0635 -16483.7441)`. It seems like the big value exceeds some limit when the image is upscaled, and Cairo goes mad. It definitely shouldn't throw "out of memory", and librsvg shouldn't panic when getting this, but things are that they are. I'm pretty sure I won't able to fix this :( But I'll try to find a way to make imgproxy not crash in such cases.

**Comment by enstyled** [comment]
Hey, Sergey,

Thank you for looking into this issue! I was also suspecting this is the source of the issue and in our case it's limited to a very small subset of files. I assume some SVG editor is generating these kind of transformations.

Ensuring imgproxy does not crash would be a good enough fix. Let me know if I can provide any more details or help.

**Comment by DarthSim** [comment]
Good news: I managed to patch the latest version of librsvg. It still fails to render your SVGs but at least it doesn't crash the entire app.
Bad news #1: The latest version of librsvg is marked as experimental, and I'd rather not use it with imgproxy. I'll try to patch the version I currently use, but this will take time without any guarantee.
Bad news #2: I'm not sure if my patch leads to memory leaks. It shouldn't but it needs some tests.

I'll try to release a patched image ASAP.

A bunch of kudos to you for catching this. This is a security hole the size of a moon crater, and it should be closed.

**Comment by DarthSim** [comment]
More good news:
- I managed to patch the currently used version of librsvg so it won't crash the entire app.
- I detected a huge memory leak, but it occurred to be caused by libvips instead of librsvg. A little patch solves the problem.

**Comment by enstyled** [comment]
Great news, Sergey, let me know when you have the release ready and I will test and report back.

Thank you for your fast reaction on this one!

**Comment by DarthSim** [comment]
Fixed in `v2.16.4`
